synapse/visibility.py

```diff
@@ -130,9 +142,10 @@ def allowed(event: EventBase) -> Optional[EventBase]:
             sender_erased=erased_senders.get(event.sender, False),
         )
 
```

Comment:
Instead of using `get_state_for_events` which gets all state for each event.

I think we can just simplify to only grab the `m.room.member` and `m.room.history_visibility` state events which is all we need to filter here.

It would be nice if we could use these simple queries instead of the recursive ones that `get_state_for_events` does but I think we do that because we have to follow the `state_group_edges` chain. `state_group` has no encompoassing guarantee based on a higher index.

```
-- Grab the state_group for the event
SELECT state_group FROM event_to_state_groups WHERE event_id = '$hbw76W6K5_LpQ4sT_yYAhly1cnC90nDBJl4AMducbhQ';

-- Grab the latest history visibility for the room
SELECT * FROM state_groups_state WHERE
    room_id = '!HBehERstyQBxyJDLfR:my.synapse.server'
    and state_group <= 1088
    and type = 'm.room.history_visibility'
ORDER BY state_group DESC
LIMIT 1;

-- Grab the latest member event for that user in the room
SELECT * FROM state_groups_state WHERE
    room_id = '!HBehERstyQBxyJDLfR:my.synapse.server'
    and state_group <= 1088
    and type = 'm.room.member'
    and state_key = '@ericgittertester:my.synapse.server'
ORDER BY state_group DESC
LIMIT 1;
```

Reply:
It looks like we already only filter down to `m.room.history_visibility` and `m.room.member` but something about the query isn't fast.

https://github.com/matrix-org/synapse/blob/e1b15f25f3ad4b45b381544ca6b3cd2caf43d25d/synapse/visibility.py#L97-L103

Reply:
See https://github.com/matrix-org/synapse/pull/14494#discussion_r1028806102 for the solution so far